<?xml version="1.0" ?>
<launch>
  <!-- LAUNCH INTERFACE -->

  <!-- The robot name is required for gazebo to spawn a model. -->
  <arg name="rightRobotName" default="right_panda"/>
  <arg name="leftRobotName" default="left_panda"/>
  <arg name="rightRobotOrigin.x" default="0"/>
  <arg name="rightRobotOrigin.y" default="-0.5"/>
  <arg name="rightRobotOrigin.z" default="0"/>
  <arg name="rightRobotOrigin.ax" default="0"/>
  <arg name="rightRobotOrigin.ay" default="0"/>
  <arg name="rightRobotOrigin.az" default="3.14159265359"/>
  <arg name="leftRobotOrigin.x" default="0"/>
  <arg name="leftRobotOrigin.y" default="0.5"/>
  <arg name="leftRobotOrigin.z" default="0"/>
  <arg name="leftRobotOrigin.ax" default="0"/>
  <arg name="leftRobotOrigin.ay" default="0"/>
  <arg name="leftRobotOrigin.az" default="3.14159265359"/>

  <arg name="surgicalFrame.x" default="-0.4"/>
  <arg name="surgicalFrame.y" default="0"/>
  <arg name="surgicalFrame.z" default="0"/>

  <arg name="target.x" default="1"/>
  <arg name="target.y" default="0"/>
  <arg name="target.z" default="0"/>
  <arg name="target2.x" default="$(arg target.x)"/>
  <arg name="target2.y" default="+0.1"/>
  <arg name="target2.z" default="$(arg target.z)"/>
  <arg name="target3.x" default="$(arg target.x)"/>
  <arg name="target3.y" default="-0.1"/>
  <arg name="target3.z" default="$(arg target.z)"/>
  <arg name="target4.x" default="$(arg target.x)"/>
  <arg name="target4.y" default="-0.2"/>
  <arg name="target4.z" default="$(arg target.z)"/>


  <arg name="sphereCenter.x" default="$(arg surgicalFrame.x)"/>
  <arg name="sphereCenter.y" default="$(arg surgicalFrame.y)"/>
  <arg name="sphereCenter.z" default="0.1"/>
  <arg name="showEnvironment" default="true"/>
  <arg name="showRightRobot" default="false"/>
  <arg name="showTarget1" default="true"/>
  <arg name="showTarget2" default="true"/>
  <arg name="showTarget3" default="true"/>
  <arg name="showTarget4" default="false"/>


  <!-- The recommended way of specifying the URDF file is to pass it in
       directly. -->
  <arg name="modelsPath" default="$(find surgical_task)/models"/>
  <arg name="environmentUrdfFile" default="$(arg modelsPath)/environment.urdf.xacro" />
  <arg name="axisUrdfFile" default="$(arg modelsPath)/axis.urdf.xacro" />
  <arg name="pandaUrdfFile" default="$(arg modelsPath)/panda_arm.urdf.xacro" />


  <arg name="rviz_config" default="$(find lwr_launch)/launch/rviz/rviz_config.rviz"/>
  <arg name="rviz_bringup" default="true"/> <!--If false, do not launch rviz-->
  <arg name="GAZEBO_GUI" default="false"/>

  
  <arg name="panda_controllers_file" default="$(find surgical_task)/config/panda_controllers.yaml"/>
  <arg name="gazebo_world_file" default="$(find surgical_simulator)/worlds/surgical_simulator_world.world"/>


  <!-- load all controller configurations to rosparam server -->
  <rosparam command="load" file="$(arg panda_controllers_file)"/>
    <!-- <node name="worldRightTransform" pkg="foot_control_xy" type="worldActorTransform" args="foot -x 2.0 -y 0.0 -z 0 -qx 0.0 -qy 0.0 -qz 0.0 -qw 1.0"/> -->

  <group if="$(arg showEnvironment)" ns="environment">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg environmentUrdfFile) origin_x:=$(arg surgicalFrame.x) origin_y:=$(arg surgicalFrame.y) origin_z:=$(arg surgicalFrame.z)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="environment_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /environment/robot_description -urdf -model environment -unpause" respawn="false" output="screen" />
  </group>


  <group ns="left_panda">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg pandaUrdfFile) robotName:=$(arg leftRobotName) origin_x:=$(arg leftRobotOrigin.x) origin_y:=$(arg leftRobotOrigin.y) origin_z:=$(arg leftRobotOrigin.z) origin_ax:=$(arg leftRobotOrigin.ax) origin_ay:=$(arg leftRobotOrigin.ay) origin_az:=$(arg leftRobotOrigin.az)"/>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /$(arg leftRobotName)/robot_description -urdf -model $(arg leftRobotName) -J $(arg leftRobotName)_joint4 -1.57 -J $(arg leftRobotName)_joint6 1.57 -unpause"/>


    <node name="state_controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="joint_state_controller joint_position_controller"/>

    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen"/>
  </group>


  <group ns="right_panda">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg pandaUrdfFile) robotName:=$(arg rightRobotName) origin_x:=$(arg rightRobotOrigin.x) origin_y:=$(arg rightRobotOrigin.y) origin_z:=$(arg rightRobotOrigin.z) origin_ax:=$(arg rightRobotOrigin.ax) origin_ay:=$(arg rightRobotOrigin.ay) origin_az:=$(arg rightRobotOrigin.az)"/>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param /$(arg rightRobotName)/robot_description -urdf -model $(arg rightRobotName) -J $(arg rightRobotName)_joint4 -1.57 -J $(arg rightRobotName)_joint6 1.57 -unpause"/>


    <node name="state_controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="joint_state_controller joint_position_controller"/>

    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" output="screen"/>
  </group>


  <group if="$(arg showTarget1)" ns="target1">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg axisUrdfFile) name:=target1 origin_x:=$(arg target.x) origin_y:=$(arg target.y) origin_z:=$(arg target.z)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="target_frame_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-x $(arg target.x) -y $(arg target.y) -z $(arg target.z) -unpause -param /target1/robot_description -urdf -model target1" respawn="false" output="screen" />
  </group>



  <group if="$(arg showTarget2)" ns="target2">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg axisUrdfFile) name:=target2 origin_x:=$(arg target.x) origin_y:=$(arg target.y) origin_z:=$(arg target.z)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="target_frame_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-x $(arg target2.x) -y $(arg target2.y) -z $(arg target2.z) -unpause -param /target2/robot_description -urdf -model target2" respawn="false" output="screen" />
  </group>



  <group if="$(arg showTarget3)" ns="target3">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg axisUrdfFile) name:=target3 origin_x:=$(arg target.x) origin_y:=$(arg target.y) origin_z:=$(arg target.z)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="target_frame_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-x $(arg target3.x) -y $(arg target3.y) -z $(arg target3.z) -unpause -param /target3/robot_description -urdf -model target3" respawn="false" output="screen" />
  </group>


  <group if="$(arg showTarget4)" ns="target4">
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(arg axisUrdfFile) name:=target4 origin_x:=$(arg target.x) origin_y:=$(arg target.y) origin_z:=$(arg target.z)"/>
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"/>
    <node name="target_frame_spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-x $(arg target4.x) -y $(arg target4.y) -z $(arg target4.z) -unpause -param /target4/robot_description -urdf -model target4" respawn="false" output="screen" />
  </group>

  <!-- GAZEBO -->
  <!-- enable/disable gui at will, the rviz listens to the simulation -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(arg gazebo_world_file)"/>
    <arg name="paused" value="true"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="$(arg GAZEBO_GUI)"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <node name="sphericalTrocarFrames" pkg="surgical_task" type="spherical_trocar_frames" args="$(arg sphereCenter.x) $(arg sphereCenter.y) $(arg sphereCenter.z) 1"/>


  <node name="pillarFrames" pkg="surgical_task" type="pillar_frames" args="1"/>

  <!-- LAUNCH RVIZ -->

  <node name="lwr_rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(arg rviz_config)"
        output="screen"
        if="$(arg rviz_bringup)"/>


</launch>
